### PR TITLE
Array.prototype.fill method implementation

### PR DESCRIPTION
**Pull request by bojan88**
Implementation for https://github.com/jasonwilliams/boa/issues/164.

### CODE REVIEW

File: src/lib/builtins/array.rs [hunk 1]
```diff
@@ -472,6 +472,43 @@ pub fn find_index(this: &Value, args: &[Value], interpreter: &mut Interpreter) -
...
+///
+/// The method fills (modifies) all the elements of an array from start index (default 0)
+/// to an end index (default array length) with a static value. It returns the modified array
+/// <https://tc39.es/ecma262/#sec-array.prototype.fill>
+pub fn fill(this: &Value, args: &[Value], _: &mut Interpreter) -> ResultValue {
+    let len: i32 = from_value(this.get_field_slice("length")).expect("Could not get argument");
+    let default_value = Gc::new(ValueData::Undefined);
+    let value = args.get(0).unwrap_or(&default_value);
+    let relateive_start: i32 = match args.get(1) {
+        Some(val) => from_value(val.clone()).expect("Could not get argument"),
```
Comment (letmutx) [inline src/lib/builtins/array.rs]: FWIW, the following also works in Firefox's js engine.
```
var a = [1, 2, 3, 1, 2, 5, 7, 11, 13, 17];
console.log(a.fill(4, "1.2", "3.4"));
```

File: src/lib/builtins/array.rs [hunk 2]
```diff
@@ -472,6 +472,54 @@ pub fn find_index(this: &Value, args: &[Value], interpreter: &mut Interpreter) -
...
+pub fn fill(this: &Value, args: &[Value], _: &mut Interpreter) -> ResultValue {
+    let len: i32 = from_value(this.get_field_slice("length")).expect("Could not get argument");
+    let default_value = Gc::new(ValueData::Undefined);
+    let value = args.get(0).unwrap_or(&default_value);
+    let relateive_start: i32 = match args.get(1) {
+        Some(val) => {
+            if val.is_string() {
+                val.to_num() as i32
+            } else {
+                val.to_int()
```
Comment (bojan88) [inline src/lib/builtins/array.rs]: @letmutx I had problem here using only `val.to_int()` - it was converting every float to zero (e.g. `'2.5'` -> `0`)
So I had to introduce this `val.to_num() as i32` conversion

Comment (letmutx) [reply src/lib/builtins/array.rs]: The following seems to work. 
```
diff --git a/src/lib/builtins/array.rs b/src/lib/builtins/array.rs
index fe3a4aa..1e6caa8 100644
--- a/src/lib/builtins/array.rs
+++ b/src/lib/builtins/array.rs
@@ -482,23 +482,11 @@ pub fn fill(this: &Value, args: &[Value], _: &mut Interpreter) -> ResultValue {
     let default_value = Gc::new(ValueData::Undefined);
     let value = args.get(0).unwrap_or(&default_value);
     let relateive_start: i32 = match args.get(1) {
-        Some(val) => {
-            if val.is_string() {
-                val.to_num() as i32
-            } else {
-                val.to_int()
-            }
-        }
+        Some(val) => val.to_num() as i32,
         None => 0,
     };
     let relative_end: i32 = match args.get(2) {
-        Some(val) => {
-            if val.is_string() {
-                val.to_num() as i32
-            } else {
-                val.to_int()
-            }
-        }
+        Some(val) => val.to_num() as i32,
         None => len,
     };
     let k = if relateive_start < 0 {
```

File: src/lib/builtins/array.rs [hunk 3]
```diff
@@ -500,10 +512,9 @@ pub fn fill(this: &Value, args: &[Value], _: &mut Interpreter) -> ResultValue {
         cmp::min(relative_end, len)
     };
 
-    while k < fin {
-        let pk = k.to_string();
+    for i in k..fin {
+        let pk = i.to_string();
         this.set_field_slice(&pk, value.clone());
```
Comment (letmutx) [inline src/lib/builtins/array.rs]: This should be `set_field`. Using `set_field_slice` does an extra allocation here.

File: src/lib/builtins/array.rs (hunk 2 above)
Comment (bojan88) [reply src/lib/builtins/array.rs]: I saw that `val.to_num() as i32` is converting `ValueData::Undefined` to a very small negative number `-2147483648`.
But seems like the tests are all succeeding. 

I'm going to add this change

Comment (letmutx) [reply src/lib/builtins/array.rs]: `undefined` should behave like `None` case. Taken from Firefox
```
var a = [1, 2, 3];
a.fill(4, undefined, undefined);
> Array(3) [ 4, 4, 4 ]
​```

Comment (bojan88) [reply src/lib/builtins/array.rs]: @letmutx thanks for discovering these issues.
I updated updated the code and added more tests. I also noticed that `NaN` is still not implemented so I commented out tests using that value

**Comment by jasonwilliams** [comment]
sorry @bojan88 there's a conflict now

**Comment by bojan88** [comment]
No problem. I resolved the conflict

File: src/lib/builtins/array.rs [hunk 4]
```diff
@@ -470,6 +470,40 @@ pub fn find_index(this: &Value, args: &[Value], interpreter: &mut Interpreter) -
...
 }
 
+/// Array.prototype.fill ( value[, start[, end]] )
+///
+/// The method fills (modifies) all the elements of an array from start index (default 0)
+/// to an end index (default array length) with a static value. It returns the modified array
+/// <https://tc39.es/ecma262/#sec-array.prototype.fill>
+pub fn fill(this: &Value, args: &[Value], _: &mut Interpreter) -> ResultValue {
+    let len: i32 = from_value(this.get_field_slice("length")).expect("Could not get argument");
+    let default_value = Gc::new(ValueData::Undefined);
```
Comment (IovoslavIovchev) [inline src/lib/builtins/array.rs]: Use the [`undefined()`](https://github.com/jasonwilliams/boa/blob/master/src/lib/builtins/value.rs#L23) function here.

File: src/lib/builtins/array.rs [hunk 5]
```diff
@@ -470,6 +470,40 @@ pub fn find_index(this: &Value, args: &[Value], interpreter: &mut Interpreter) -
...
+    let value = args.get(0).unwrap_or(&default_value);
+    let relative_start = args.get(1).unwrap_or(&default_value).to_num() as i32;
+    let relative_end_val = args.get(2).unwrap_or(&default_value);
+    let relative_end = if relative_end_val.is_undefined() {
+        len
+    } else {
+        relative_end_val.to_num() as i32
+    };
+    let k = if relative_start < 0 {
```
Comment (IovoslavIovchev) [inline src/lib/builtins/array.rs]: Could you give a more descriptive name than `k`?

File: src/lib/builtins/array.rs [hunk 6]
```diff
@@ -573,6 +607,7 @@ pub fn create_constructor(global: &Value) -> Value {
     array_prototype.set_field_slice("includes", includes_func);
     array_prototype.set_field_slice("indexOf", index_of_func);
     array_prototype.set_field_slice("lastIndexOf", last_index_of_func);
+    array_prototype.set_field_slice("fill", to_value(fill as NativeFunctionData));
```
Comment (IovoslavIovchev) [inline src/lib/builtins/array.rs]: You need to set the `length` property of a function.

Comment (bojan88) [reply src/lib/builtins/array.rs]: @IovoslavIovchev can you please be more specific? I'm afraid I don't understand this comment

File: src/lib/builtins/array.rs (hunk 5 above)
Comment (bojan88) [reply src/lib/builtins/array.rs]: Can you please propose any better name? What do you think about `start`?
The value is named `k` both in tc39 spec and in MDN polyfill example

Comment (IovoslavIovchev) [reply src/lib/builtins/array.rs]: Yeah, `start` is okay.

File: src/lib/builtins/array.rs (hunk 6 above)
Comment (IovoslavIovchev) [reply src/lib/builtins/array.rs]: `Function`s have their `length` property set to the number of required parameters they have (e.g. those with no default value).

You can take a look at how it's currently done [here](https://github.com/bojan88/boa/blob/1728795ff784ffabe3ad88369a9bd2fe7a5b4f3a/src/lib/builtins/array.rs#L596).

Comment (bojan88) [reply src/lib/builtins/array.rs]: Thanks @IovoslavIovchev. I didn't know that `length` property existed on this function/object.

File: src/lib/builtins/array.rs [hunk 7]
```diff
@@ -936,6 +973,109 @@ mod tests {
...
+        // make sure the array is modified
+        assert_eq!(forward(&mut engine, "a.join()"), String::from("4,4,4"));
+
+        forward(&mut engine, "a = [1, 2, 3];");
+        assert_eq!(
+            forward(&mut engine, "a.fill(4, '1').join()"),
+            String::from("1,4,4")
+        );
+
+        forward(&mut engine, "a = [1, 2, 3];");
```
Comment (IovoslavIovchev) [inline src/lib/builtins/array.rs]: It seems you added the same test twice. Remove this one.

**Pull request merged**